### ISSUE TITLE
Intermittent column type resolution when using Quarkus

### ISSUE THREAD

**Issue by YassinHajaj** [vladmihalcea/hypersistence-utils#599]
**Setup details**

* Quarkus 2.16.4.Final
* Hypersistence 3.2.0
* Postgresql 13.x

**Problem description**

For 99% of our queries, no issue is to be found.
This happens because we noticed that the column was treated as a json column (using JsonTypeDescriptor & JsonBinaryType).

But sometimes, randomly, at startup or during the life of an app, it crashes with the following error.
It's clear it doesn't use the JsonTypeDescriptor and its JsonBinaryType anymore, but it considers the column now as being a straight array.

```
Caused by: java.lang.IllegalArgumentException: org.postgresql.util.PSQLException: No results were returned by the query.
	... 122 more
Caused by: org.postgresql.util.PSQLException: No results were returned by the query.
	at org.postgresql.jdbc.TypeInfoCache.getArrayDelimiter(TypeInfoCache.java:614)
	at org.postgresql.jdbc.PgArray.buildArrayList(PgArray.java:298)
	at org.postgresql.jdbc.PgArray.getArrayImpl(PgArray.java:157)
	at org.postgresql.jdbc.PgArray.getArray(PgArray.java:111)
	at io.hypersistence.utils.hibernate.type.array.internal.AbstractArrayTypeDescriptor.wrap(AbstractArrayTypeDescriptor.java:94)
```

We've noticed, by following the stacktrace that a query occured on Postgres to get the delimiter for the resolved type (jsonb). Problem is, this is not an array type, just a regular jsonb. Even if we persist arrays in it.

The query that postgres executes is the following
```
SELECT *
FROM pg_catalog.pg_type t, pg_catalog.pg_type e 
WHERE t.oid = 3802 and t.typelem = e.oid;
```

But as our type is not an _actual_ array, we don't need this query to be executed.
We found the actual array type for jsonb to have OID 3807.

Any idea ?

**Comment by vladmihalcea** [comment]
It's not clear what the problem is. If it's for the entity mapping, I don't see how it could choose an ArrayType when you explicitly tell Hibernate to use the JsonType.

So, is it for mapping the result set?

Does this problem occur when you try to bind a Json to a DML statement?

**Comment by DeMol-EE** [comment]
Hi @vladmihalcea, I’ll try to clarify (I’m working with @YassinHajaj on this project).

First of all, we have put `@TypeDef(name = "jsonb", typeClass = JsonBinaryType.class)` on the entity class to register the type. Then, in the entity, we have a few fields of type `Set<Foo>` (where `Foo` is not just a boxed type, though I doubt that’s relevant) which we have annotated with `@Type(type = "jsonb")`. We’ve also placed `columnDefinition = "jsonb"` on the column annotation of each of those fields.
This describes the mapping to a table with jsonb columns in our postgres db. There’s nothing special about it, perhaps maybe the fact we’re using jsonb at all.

Now the problem is that our Quarkus app sometimes crashes with the stack trace that Yassin has provided above. Sometimes this happens at startup, sometimes it happens after running steadily for hours. (In case it might be relevant: we’re running Quarkus in JVM mode, we’re not using native compilation.)

What’s puzzling about this is that the code works fine for the vast majority of the time. So for hours or even days our application might be processing thousands of requests that read and write to these jsonb fields without issue, but then it suddenly fails (and keeps failing). This makes it really hard to reproduce and investigate. When it happens to someone in the development team, they typically just restart the application or re-run the tests (it happens sometimes when running `@QuarkusTest`s as well, as this starts quarkus). When it happens in our QA environment, we need to deploy a new pod.

Yassin has conducted a thorough investigation with the debugger and has found a possible cause for the issue. He has found that in most cases (when the code is working fine), the `JsonTypeDescriptor` is detected by Hibernate (provided by your library) but that in the rare cases where it all breaks, `AbstractArrayTypeDescriptor` is detected instead (for the same field on the same entity with the same source code).

Given the seemingly random nature of the problem, we were thinking in the direction of a race condition, a classloader issue or some other source of randomness (accessing the first element of a Set, perhaps), but we are simply not sure because we can not pinpoint the source of the issue further. We can also not create a reproducer project. We could put the minimal amount of code that triggers the issue in a small project, but we can not reliably reproduce the problem. I would say it happens on average once per week, while doing full-time development. It has happened when running tests, when running locally, during live reloading in quarkus’ dev mode, in our pipeline (also when running the tests) and in our environments, both at start up (immediately broken when a new pod starts) and suddenly after hours of service. It’s a surprise each time.

With the debugger, we have seen that Hibernate inspects the entities and dynamically creates metamodels (in `AbstractEntityPersister`) when the Quarkus application launches. It’s also here that we’ve seen that `getPropertyTypes` returns `JsonBinaryType`, but that it’s accompanying `javaTypeDescriptor` is not always the same. As stated above: when it’s `JsonTypeDescriptor`, the app runs fine, but sometimes it’s `AbstractArrayTypeDecription` and then the app won’t start. In those cases, it runs the query that Yassin provided above to select from pg_catalog but with the wrong oid, which returns no rows (hence the error in the stack trace).
We theorise that the result of this inspection is refreshed at some point(s) during the application’s lifetime, purely to have an explanation for why it sometimes happens that our application starts to fail after hours of service.

It’s been months and the issue seems to be occurring more and more frequently (though still rarely). Originally, I would say it happened once a month and now it’s sometimes daily on our QA environment. We’re not in production yet but we are nearing our release date and we would like to see the problem resolved before then.
We’re getting quite desperate and are currently considering removing all usage of jsonb, or migrating to Quarkus 3 to get Hibernate 6 which supports json out of the box (hoping that the issue disappears with that). We have tried updating our quarkus version, your library (we originally had hibernate-types), the jdbc driver but all to no avail.
Given the information we are able to provide, we were hoping you might be able to find the root cause of the issue.
